src/index.ts

```diff
@@ -379,38 +388,111 @@ class ThreadPool {
+    while ((this.taskQueue.size > 0 || this.skipQueue.length > 0)) {
       // The skipQueue will have tasks that we previously shifted off
       // the task queue but had to skip over... we have to make sure
       // we drain that before we drain the taskQueue.
       const taskInfo = this.skipQueue.shift() ||
                        this.taskQueue.shift() as TaskInfo;
-      // If the task has an abortSignal and the worker has any other
-      // tasks, we cannot distribute the task to it. Skip for now.
-      if (taskInfo.abortSignal && workerInfo.taskInfos.size > 0) {
-        this.skipQueue.push(taskInfo);
-        break;
+
+      if (workers == null) {
+        workers = [...this.workers].map(workerInfo => workerInfo.interface);
       }
-      const now = performance.now();
-      this.waitTime?.record(toHistogramIntegerNano(now - taskInfo.created));
-      taskInfo.started = now;
-      workerInfo.postTask(taskInfo);
-      this._maybeDrain();
-      return;
+
+      // console.log('distributing task', taskInfo.interface);
+      const distributed = this._distributeTask(taskInfo, workers);
+
+      // If task was distributed, we should continue to distribute more tasks
+      if (distributed) {
+        continue;
+        // If balancer states that pool is busy, we should stop trying to distribute tasks
+      } else { break; }
     }
 
-    if (workerInfo.taskInfos.size === 0 &&
+    // If more workers than minThreads, we can remove idle workers
+    if (workerInfo.currentUsage() === 0 &&
         this.workers.size > this.options.minThreads) {
       workerInfo.idleTimeout = setTimeout(() => {
-        assert.strictEqual(workerInfo.taskInfos.size, 0);
+        assert.strictEqual(workerInfo.currentUsage(), 0);
         if (this.workers.size > this.options.minThreads) {
           this._removeWorker(workerInfo);
         }
       }, this.options.idleTimeout).unref();
     }
   }
 
+  _distributeTask (task: TaskInfo, workers: PiscinaWorker[]): boolean {
+    const balancerResult = this.balancer(task.interface, workers);
+    // TODO: we need to verify if the task is aborted already or not
+    // otherwise we might be distributing aborted tasks to workers
+    if (task.aborted) return false;
+    // console.log(task.interface, balancerResult);
+
+    if (balancerResult.candidate != null) {
+      const now = performance.now();
+      this.waitTime?.record(toHistogramIntegerNano(now - task.created));
+      task.started = now;
+      balancerResult.candidate[kWorkerData].postTask(task);
+      this._maybeDrain();
+      // If candidate, let's try to distribute more tasks
+      return true;
+    }
+
+    switch (balancerResult.command) {
+      case 0: { // busy
+        if (task.abortSignal) {
+          // console.log('into skipqueue');
+          this.skipQueue.push(task);
+        } else {
+          this.taskQueue.push(task);
+        }
+
+        return false;
+      }
+      case 1: { // add
```

Comment:
It's not the LB code that must take such a decision. It's only responsibility is to pick up a worker given custom criteria among the ones available if any. Worker creation is orthogonal to load balancing (and done before invoking the LB). 
Furthermore, if will make writing custom LB more complex and error prone than it should.

Reply:
Hmm, this is an interesting point; maybe I saw it in another way.
My initial idea was to help the balancer hint the pool wether to instantiate a new worker or not; but overall, if the balancer decides not to pick one, the pool can make the further decision of spawning and enqueuing.

That can make it cleaner.

Reply:
Since you're targeting a public API, I really think it should be as simple as possible, like the custom queue one. 

Regarding the worker creation usage, you can add optional tunables with sensible defaults to the LB API to allow to give hint to the pool behavior, such as the direct usage of any newly created worker instead running the LB algo with it added.  

Something like: 
```ts
{ 
  useCreatedWorker: boolean,
  ...
 }
```

Reply:
That would be definitely something interesting to explore, but maybe we can start simply and state that if the balancer returns `null`, Piscina will attempt to enqueue the task or start a new worker.

Further we can explore how a balancer and scheduler can collaborate together; but maybe for simplicity just keep it as is

Reply:
I do not think it's a good idea to expect the worker creation decision logic to be part of the LB code. It will end by having the same code in all LBs: 

Is there a worker available? -> Yes -> Run the LB algo | (-> No -> Is pool full -> Yes -> End | -> No -> Ask for worker creation) 
The LB code will need to be called twice.  

That logic can be factored out the LB code and makes LB implementation easier.

Reply:
> The LB code will need to be called twice.

Hmm, this is interesting. It will require to split the handling of a new worker from when a worker finishes a task, which is currently coupled. That will need to be handled differently.

In that way, upon creation we deliver the first task in the queue; but I imagined the LB, although called twice, might have some benefits if the algorithm is based on some affinity. 

>The extra work worth it, it will evolve since it's not like a queue API, it's more like a plugin API.

Yeah, that's the intention. Thanks for the callout

Reply:
> > The LB code will need to be called twice.
> 
> Hmm, this is interesting. It will require to split the handling of a new worker from when a worker finishes a task, which is currently coupled. That will need to be handled differently.

I guess the code entanglement for worker creation logic and load balancing logic can be part of another PR. I would do it prior to LB API integration. If I remember correctly, in piscina, at task end (message channel response):
- lock is taken on the shared tasks queue 
- LB is run to find an available worker with a minimum of executing tasks, if none available, worker is created if pool not full
- lock is released

I'm not sure about the locking, it's probably taken only if an eligible worker is found. Whatever, testing for workers availability and readyness and create one if needed logic can be factored out and run once in the current code as an optimization. And will help for that PR.



> 
> In that way, upon creation we deliver the first task in the queue; but I imagined the LB, although called twice, might have some benefits if the algorithm is based on some affinity.
> 
> > The extra work worth it, it will evolve since it's not like a queue API, it's more like a plugin API.
> 
> Yeah, that's the intention. Thanks for the callout

Reply:
Yeah, it is not really locking (but mostly a side-effect). Piscina infers that if you already have a task queue that hasn't been emptied yet, new tasks are enqueued so this ensures the tasks are solved in the same way they entered (unless a custom task queue has been set).

Nevertheless, the thread is stopped until a further task is picked.

That sounds good, think that gonna make the implementation easier and move forward.